File: packages/service-auth/src/delegation.ts
import { FetchLike, ServiceAuthError } from './types';
import type { ServiceAuthClient } from './client';

export interface DelegationExchangeOptions {
  baseUrl: string;
  serviceAuth: ServiceAuthClient;
  fetch?: FetchLike;
  timeoutMs?: number;
}

export interface DelegationRequest {
  subjectToken: string;
  audience: `service:${string}`;
  scopes: string[];
}

export interface DelegationToken {
  accessToken: string;
  tokenType: 'Bearer';
  expiresIn: number;
  scope: string;
  subject: string;
  audience: string;
  actor: { sub: string; previous?: unknown };
}

export interface DelegationClient {
  exchange(request: DelegationRequest): Promise<DelegationToken>;
}

export function createDelegationClient(options: DelegationExchangeOptions): DelegationClient {
  if (!options.baseUrl || !options.serviceAuth) {
    throw new ServiceAuthError('MISCONFIGURED', 'Delegation client requires baseUrl and serviceAuth.', 500);
  }
  const fetcher = options.fetch ?? (globalThis.fetch?.bind(globalThis) as unknown as FetchLike);
  if (!fetcher) throw new ServiceAuthError('MISCONFIGURED', 'No fetch implementation available.', 500);
  const endpoint = `${options.baseUrl.replace(/\/+$/, '')}/api/v1/security/tokens/exchange`;

  return {
    async exchange(request) {
      if (!request.subjectToken || !request.audience || request.scopes.length === 0) {
        throw new ServiceAuthError('MISCONFIGURED', 'subjectToken, audience and scopes are required.', 400);
      }
      const actorToken = await options.serviceAuth.getToken();
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), options.timeoutMs ?? 5000);
      try {
        const response = await fetcher(endpoint, {
          method: 'POST',
          headers: { authorization: `Bearer ${actorToken}`, 'content-type': 'application/json' },
          body: JSON.stringify({
            subjectToken: request.subjectToken,
            audience: request.audience,
            scope: [...new Set(request.scopes)].join(' '),
          }),
          signal: controller.signal,
        });
        if (!response.ok) {
          throw new ServiceAuthError(
            response.status === 403 ? 'FORBIDDEN' : 'TOKEN_REQUEST_FAILED',
            `Delegation exchange returned ${response.status}.`,
            response.status,
          );
        }
        const body = (await response.json()) as Partial<DelegationToken>;
        if (!body.accessToken || !body.actor?.sub || !body.audience) {
          throw new ServiceAuthError('MALFORMED_RESPONSE', 'Delegation response is incomplete.', 502);
        }
        return body as DelegationToken;
      } finally {
        clearTimeout(timeout);
      }
    },
  };
}

File: packages/service-auth-py/src/fuzefront_service_auth/delegation.py
"""Constrained on-behalf-of token exchange against FuzeFront Security."""

from __future__ import annotations

import json
from dataclasses import dataclass

from ._http import HttpPost, default_http_post
from .client import ServiceAuthClient
from .exceptions import TokenRequestError


@dataclass(frozen=True)
class DelegationToken:
    access_token: str
    expires_in: int
    scope: str
    subject: str
    audience: str
    actor: dict

    @property
    def authorization_header(self) -> str:
        return f"Bearer {self.access_token}"


class DelegationClient:
    def __init__(
        self,
        base_url: str,
        service_auth: ServiceAuthClient,
        *,
        timeout: float = 10.0,
        http_post: HttpPost | None = None,
    ) -> None:
        if not base_url or service_auth is None:
            raise TokenRequestError("DelegationClient requires base_url and service_auth", code="MISCONFIGURED", status=500)
        self._base_url = base_url.rstrip("/")
        self._service_auth = service_auth
        self._timeout = timeout
        self._http_post = http_post or default_http_post

    def exchange(self, subject_token: str, audience: str, scopes: list[str]) -> DelegationToken:
        if not subject_token or not audience.startswith("service:") or not scopes:
            raise TokenRequestError("subject_token, service audience and scopes are required", status=400)
        actor = self._service_auth.get_token()
        payload = {
            "subjectToken": subject_token,
            "audience": audience,
            "scope": " ".join(dict.fromkeys(scopes)),
        }
        status, body = self._http_post(
            f"{self._base_url}/api/v1/security/tokens/exchange",
            payload,
            self._timeout,
            headers={"Authorization": actor.authorization_header},
        )
        if status != 200:
            raise TokenRequestError(f"delegation exchange returned HTTP {status}", status=status)
        try:
            data = json.loads(body)
            return DelegationToken(
                access_token=data["accessToken"],
                expires_in=int(data["expiresIn"]),
                scope=data["scope"],
                subject=data["subject"],
                audience=data["audience"],
                actor=data["actor"],
            )
        except (KeyError, TypeError, ValueError, json.JSONDecodeError) as error:
            raise TokenRequestError("malformed delegation response", code="MALFORMED_RESPONSE", status=502) from error
